File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/02．机制单位生命周期模板.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____01_FF0E_53EF_653B_51FB_673A_5236_5355_4F4D = require("系统.03．技能系统.00．技能模板+函数.04．机制组件.05．机制单位.01．可攻击机制单位")
local _____521B_5EFA_53EF_653B_51FB_673A_5236_5355_4F4D = ____01_FF0E_53EF_653B_51FB_673A_5236_5355_4F4D["创建可攻击机制单位"]
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local addDelayedCallback = ____require_result_0.addDelayedCallback
local removeDelayedCallback = ____require_result_0.removeDelayedCallback
local _____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0 = __TS__Class()
_____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0.name = "机制单位生命周期实现"
function _____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0.prototype.____constructor(self, _____57FA_7840_5B9E_4F8B, _____53C2_6570)
    self["超时回调ID"] = 0
    self["已结束"] = false
    self["基础实例"] = _____57FA_7840_5B9E_4F8B
    self["单位"] = _____57FA_7840_5B9E_4F8B["单位"]
    self["参数"] = _____53C2_6570
    if _____53C2_6570["超时秒"] ~= nil and _____53C2_6570["超时秒"] > 0 then
        local ____self = self
        self["超时回调ID"] = addDelayedCallback(
            _____53C2_6570["超时秒"] * 1000,
            function()
                ____self["处理超时"](____self)
            end
        )
    end
end
_____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0.prototype["是否存活"] = function(self)
    return not self["已结束"] and self["基础实例"]["是否存活"]()
end
_____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0.prototype["销毁"] = function(self, _____539F_56E0)
    if _____539F_56E0 == nil then
        _____539F_56E0 = "手动销毁"
    end
    if self["已结束"] then
        return
    end
    self["结束"](self, _____539F_56E0)
    self["基础实例"]["销毁"]()
end
_____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0.prototype["处理死亡"] = function(self, _____51FB_6740_8005)
    if self["已结束"] then
        return
    end
    if self["参数"]["on被摧毁"] ~= nil then
        self["参数"]["on被摧毁"](self, _____51FB_6740_8005)
    end
    self["结束"](self, "被摧毁")
end
_____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0.prototype["处理超时"] = function(self)
    if self["已结束"] then
        return
    end
    if self["参数"]["on超时"] ~= nil then
        self["参数"]["on超时"](self)
    end
    if self["参数"]["超时后销毁"] ~= false then
        self["销毁"](self, "超时")
    else
        self["结束"](self, "超时")
    end
end
_____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0.prototype["结束"] = function(self, _____539F_56E0)
    if self["已结束"] then
        return
    end
    self["已结束"] = true
    if self["超时回调ID"] ~= 0 then
        removeDelayedCallback(self["超时回调ID"])
        self["超时回调ID"] = 0
    end
    if self["参数"]["on结束"] ~= nil then
        self["参数"]["on结束"](self, _____539F_56E0)
    end
end
____exports["创建机制单位生命周期"] = function(_____53C2_6570)
    local _____5B9E_4F8B
    local _____57FA_7840_5B9E_4F8B = _____521B_5EFA_53EF_653B_51FB_673A_5236_5355_4F4D(__TS__ObjectAssign(
        {},
        _____53C2_6570,
        {
            ["on死亡"] = function(_____5355_4F4D, _____51FB_6740_8005)
                if _____53C2_6570["on死亡"] ~= nil then
                    _____53C2_6570["on死亡"](_____5355_4F4D, _____51FB_6740_8005)
                end
                if _____5B9E_4F8B ~= nil then
                    _____5B9E_4F8B["处理死亡"](_____5B9E_4F8B, _____51FB_6740_8005)
                end
            end,
            ["on销毁"] = function(_____5355_4F4D)
                if _____53C2_6570["on销毁"] ~= nil then
                    _____53C2_6570["on销毁"](_____5355_4F4D)
                end
            end
        }
    ))
    if _____57FA_7840_5B9E_4F8B == nil then
        return nil
    end
    _____5B9E_4F8B = __TS__New(_____673A_5236_5355_4F4D_751F_547D_5468_671F_5B9E_73B0, _____57FA_7840_5B9E_4F8B, _____53C2_6570)
    if _____53C2_6570["清理"] ~= nil then
        local ____self_1 = _____53C2_6570["清理"]
        ____self_1["登记清理"](
            ____self_1,
            _____53C2_6570["名称"],
            function()
                if _____5B9E_4F8B ~= nil then
                    _____5B9E_4F8B["销毁"](_____5B9E_4F8B)
                end
            end
        )
    end
    if _____53C2_6570["on创建"] ~= nil then
        _____53C2_6570["on创建"](_____5B9E_4F8B)
    end
    return _____5B9E_4F8B
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/02．机制单位生命周期模板.ts
/** @noSelfInFile */

import { 创建可攻击机制单位, 可攻击机制单位参数, 可攻击机制单位实例 } from "../05．机制单位/01．可攻击机制单位";
import type { 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";

const { addDelayedCallback, removeDelayedCallback } = require("系统.00．核心系统.05．中心计时器") as {
  addDelayedCallback: (this: void, delayMs: number, callback: (this: void) => void) => number;
  removeDelayedCallback: (this: void, id: number) => void;
};

export interface 机制单位生命周期参数 extends 可攻击机制单位参数 {
  清理?: 机制清理篮子;
  名称: string;
  超时秒?: number;
  超时后销毁?: boolean;
  on创建?: (this: void, 实例: 机制单位生命周期实例) => void;
  on被摧毁?: (this: void, 实例: 机制单位生命周期实例, 击杀者: any) => void;
  on超时?: (this: void, 实例: 机制单位生命周期实例) => void;
  on结束?: (this: void, 实例: 机制单位生命周期实例, 原因: 机制单位生命周期结束原因) => void;
}

export type 机制单位生命周期结束原因 = "被摧毁" | "超时" | "手动销毁";

export interface 机制单位生命周期实例 {
  readonly 单位: any;
  readonly 基础实例: 可攻击机制单位实例;
  是否存活(): boolean;
  销毁(原因?: 机制单位生命周期结束原因): void;
}

class 机制单位生命周期实现 implements 机制单位生命周期实例 {
  readonly 单位: any;
  readonly 基础实例: 可攻击机制单位实例;
  private 参数: 机制单位生命周期参数;
  private 超时回调ID = 0;
  private 已结束 = false;

  constructor(基础实例: 可攻击机制单位实例, 参数: 机制单位生命周期参数) {
    this.基础实例 = 基础实例;
    this.单位 = 基础实例.单位;
    this.参数 = 参数;
    if (参数.超时秒 != null && 参数.超时秒 > 0) {
      const self = this;
      this.超时回调ID = addDelayedCallback(参数.超时秒 * 1000, function 机制单位生命周期超时(this: void): void {
        self.处理超时();
      });
    }
  }

  是否存活(): boolean {
    return !this.已结束 && this.基础实例.是否存活();
  }

  销毁(原因: 机制单位生命周期结束原因 = "手动销毁"): void {
    if (this.已结束) return;
    this.结束(原因);
    this.基础实例.销毁();
  }

  处理死亡(击杀者: any): void {
    if (this.已结束) return;
    if (this.参数.on被摧毁 != null) this.参数.on被摧毁(this, 击杀者);
    this.结束("被摧毁");
  }

  private 处理超时(): void {
    if (this.已结束) return;
    if (this.参数.on超时 != null) this.参数.on超时(this);
    if (this.参数.超时后销毁 !== false) {
      this.销毁("超时");
    } else {
      this.结束("超时");
    }
  }

  private 结束(原因: 机制单位生命周期结束原因): void {
    if (this.已结束) return;
    this.已结束 = true;
    if (this.超时回调ID !== 0) {
      removeDelayedCallback(this.超时回调ID);
      this.超时回调ID = 0;
    }
    if (this.参数.on结束 != null) this.参数.on结束(this, 原因);
  }
}

export function 创建机制单位生命周期(this: void, 参数: 机制单位生命周期参数): 机制单位生命周期实例 | undefined {
  let 实例: 机制单位生命周期实现 | undefined;
  const 基础实例 = 创建可攻击机制单位({
    ...参数,
    on死亡: function 机制单位生命周期死亡(this: void, 单位: any, 击杀者: any): void {
      if (参数.on死亡 != null) 参数.on死亡(单位, 击杀者);
      if (实例 != null) 实例.处理死亡(击杀者);
    },
    on销毁: function 机制单位生命周期销毁(this: void, 单位: any): void {
      if (参数.on销毁 != null) 参数.on销毁(单位);
    },
  });
  if (基础实例 == null) return undefined;
  实例 = new 机制单位生命周期实现(基础实例, 参数);
  if (参数.清理 != null) {
    参数.清理.登记清理(参数.名称, function 机制单位生命周期清理(this: void): void {
      if (实例 != null) 实例.销毁();
    });
  }
  if (参数.on创建 != null) 参数.on创建(实例);
  return 实例;
}
